Fix relpath handling in file.recurse

ScoreUnder opened the pull request:
### What does this PR do?

From commit message:

> # Fix relpath handling in file.recurse
>
> * Ensure that the "maxdepth" checks are performed on all relative paths
>   encountered during the file.recurse call, including symlink and
>   directory paths.
> * Ensure that directory transversal does not occur.
> 
> No attempt is made to address the issue of master nodes sending more
> files/directories than requested via the cp.list_master family of
> functions.

### What issues does this PR fix or reference?

#39915 

### Previous Behavior

* Symlink handling:
  * No check for directory transversal (use of the `..` path component)
* File handling:
  * Directory transversal check skips directories whose names begin with `..`.
* Empty directory handling:
  * No check for directory transversal
  * No check for max depth being exceeded

### New Behavior

* All three of the above now check the same things:
  * Directory transversal (`..` path component) must not be present anywhere in the path
  * Path components containing `..` are allowed, as long as they are not exactly `..`
  * Max depth is always checked

### Tests written?

No

This fixes the issue with manual testing, but manual testing is not very comprehensive ;)

ScoreUnder committed:
Fix relpath handling in file.recurse

* Ensure that the "maxdepth" checks are performed on all relative paths
  encountered during the file.recurse call, including symlink and
  directory paths.
* Ensure that directory transversal does not occur.

No attempt is made to address the issue of master nodes sending more
files/directories than requested via the cp.list_master family of
functions.

Fix #39915

ScoreUnder committed:
Replace split with strip/count

ScoreUnder commented:
Added a last-minute bikeshed of a commit :) I think it makes that part of the code significantly clearer.

terminalmage commented on salt/states/file.py:
```diff
@@ -352,6 +352,33 @@ def _check_user(user, group):
+    '''
+    # Check relpath surrounded by slashes, so that `..` can be caught as
+    # a path component at the start, end, and in the middle of the path.
+    if '/../' in '/{0}/'.format(relpath):
```
This will not work in Windows, since ``os.path.relpath()`` uses a backslash as a path separator on that platform. ``/`` should be replaced with ``os.sep``.

Also, can you explain the reasoning behind putting a slash after the ``..``? Your comment above is not very clear. ``os.path.relpath()`` doesn't include these for directories. However, ``file.recurse`` should just operate on file paths, and create directories as needed. Maybe that's the reason?

ScoreUnder replied:
The slash after (and before) the `..` is to ensure that it is a whole path component. So that this picks up `../some_path` too (and other cases), the whole path is first surrounded in slashes. This is easier than a test for `(^|/)..($|/)` but otherwise the same.

The original test used `startswith("..")` which has the side effect of rejecting too many paths (any files that start with `..`).

Regarding Windows compatibility, I'm not sure how to proceed. I don't have any Windows machine to test on (and python doesn't seem to work on Wine). All I know is that Windows accepts both `/` and `\` as a path separator, meaning that `os.sep` must be incorrect (because it has to pick only one), and I don't know whether or not `os.path.relpath` normalises all path separators.

For clarity, my main worry is because these paths are identical on Windows:

```
../path/file
..\path\file
..\path/file
../path\file
```

Approaches and their problems:

* Deal with `os.sep`
  * Forward-slashes break despite being valid
  * Does relpath normalise to one or the other?
* Split parts of the function into "windows" and "everyone else" versions
  * "everyone else" being the `os.sep` version
* Replace backslashes with forward slashes
  * Backslashes are a valid filename character on other operating systems.
  * Is there an OS which uses backslashes which doesn't support forward slashes?
  * Probably a bad idea all around
* Write a different `relpath` that works more sanely (i.e. deals with forward-slash-separated paths only) and transform into an `os.sep`-separated path at the 
  * Might not be necessary depending on how Windows' `relpath` behaves
* Use `os.sep` anyway and let windows users complain when it doesn't work
  * I like this idea :^)

I'm not sure which option to take here.
It's also worth pointing out that the original code made the same assumptions about platform and path format.

ScoreUnder replied:
Just come across the `posixpath`/`ntpath` modules. This will simplify things.

terminalmage replied:
Sorry, please disregard. I overlooked the fact that the paths are from cp.list_master so they are ``salt://`` paths. Sorry for the confusion on my part.

ScoreUnder replied:
I did end up looking into the differing relpath implementations though, and there is definitely some cause for concern. I'm considering using `posixpath` to deal with the salt master paths, and that just leaves the concern of what to do with the paths when actually copying to the filesystem (i.e. whether or not they'll need transformation). I'll give that an attempt today.

terminalmage commented on salt/states/file.py:
```diff
@@ -352,6 +352,33 @@ def _check_user(user, group):
+    # Check that the relative path's depth does not exceed maxdepth
+    if maxdepth is not None:
+        # Since paths are all master, just use POSIX separator
+        path_depth = relpath.strip('/').count('/')
```
You can't assume this. We're working on getting the master running on Windows, so you must use ``os.sep`` here.

ScoreUnder replied:
Will the Windows master not return slash-separated paths? If the master starts returning different data to minions depending on its OS, that is a recipe for trouble.

terminalmage replied:
You are right, I misread the surrounding code and did not think it was operating on ``salt://`` paths.

terminalmage reviewed (changes requested):
Some changes to be made here, and also a number of issues caught by our lint check.

ScoreUnder committed:
Maintain separation between master and minion paths

Paths received through the master's API are in POSIX-compatible format,
while paths on the minion are in the OS's native format.
Ensure that the correct set of functions (posixpath vs os.path) is used
on each.

ScoreUnder committed:
Remove extra parens

Pylint doesn't like them.

ScoreUnder commented:
Attempted to split the two types of paths. OS paths are handled with `os.path`, and master paths with `posixpath`. I also replaced a couple of instances of `os.sep` with `os.path.sep` so that they show up when grepping `os.path`.

Lacking a Windows machine to test on I took the somewhat questionable approach of `os.path = ntpath` and tested the modified functions in isolation.

cachedout commented:
Go Go Jenkins!

cachedout commented:
@terminalmage I think this may be ready for you to have another look.

terminalmage reviewed (approved):
:+1:

cachedout commented:
Go Go Jenkins!

cachedout commented:
Linted locally and it passed. Merging.

seanjnkns commented:
Just updated to 2016.11.5, released today and doesn't appear this PR made it.  Was it not backported for some reason?

Salt Version:
           Salt: 2016.11.5
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: 5.6.0
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: 1.4.3
       M2Crypto: 0.21.1
           Mako: 0.8.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: 1.2.5
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.3.1611 Core
        machine: x86_64
        release: 4.9.28-1.el7.centos.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core

cachedout commented:
@seanjnkns It was not. Was there a request for that to happen? I don't think I saw it.

seanjnkns commented:
@gtmanfred responded on issue #41333 that it wasn't and it would be backported.  I was hoping that it would have been done automatically as part of seeing it affects 2016.3 and 2016.11 is based off 2016.3, so one would have thought it would have automatically been ported to the newer release.  Sadly no.  If/when the patch is backported, even if a new version hasn't been cut, I'll likely need to apply it on top of 2016.11.5 as it's causing issues for user management and file.recurse of their home directory files as users like matt and mattr collide (ie, matt clobbers mattr due to the wildcarding nature of file.recurse pre-patch)

gtmanfred commented:
Here is our doc which describes how we do forward merging and which branch a patch should be applied to https://docs.saltstack.com/en/latest/topics/development/contributing.html#which-salt-branch

seanjnkns commented:
Thanks @gtmanfred, but unless I'm missing something here, #39915 was opened against 2016.3 but the PR was submitted to develop and should have had a backport tag on it to begin with, or ideally, the PR should have been submitted against the oldest release, which at this time is 2016.3, and then forward-ported to 2016.11 and develop.  Given this didn't happen, it makes me concerned there's likely a decent amount of bug/documentation disparity between the 3 main branches, 2016.3, 2016.11, and develop.  In my free time, I'll try and come up with some method to check all three branches and try and establish which PRs weren't properly merged into the respective branches as they should have.

gtmanfred commented:
Yes, the issue was opened against 2016.3, but the pr was opened to fix this in develop.

saltstack:develop from ScoreUnder:39915_fix_relpath_handling

We do not by default backport patches if the person submitting them puts them against develop, unless it is specifically requested to be backported, because we have a policy of only doing forward merges on a regular basis.

rallytime commented:
@gtmanfred and @seanjnkns This PR will not backport cleanly to 2016.11, as it edits code that is not present in the 2016.11 branch. Therefore, I won't be back-porting this change. The 2016.11 branch will need its own fix for the issue at hand.

Note, however, that this PR will be present in the upcoming `nitrogen` feature release.
